File: clinic-referral-dashboard/node/src/evageneClient.ts
/**
 * Thin client for the Evagene REST endpoints the dashboard needs.
 *
 * One method per endpoint.  All I/O goes through the injected HttpGateway
 * so tests can replace it with a fake.  Errors carry enough context for
 * the caller to decide whether to render a degraded card or abort.
 */

import type { HttpGateway } from './httpGateway.js';

const HTTP_OK_LOWER = 200;
const HTTP_OK_UPPER = 300;

export class EvageneApiError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'EvageneApiError';
  }
}

export interface PedigreeSummary {
  readonly id: string;
  readonly displayName: string;
}

export interface EvageneClientOptions {
  readonly baseUrl: string;
  readonly apiKey: string;
  readonly http: HttpGateway;
}

export interface EvageneApi {
  fetchEmbedSvg(pedigreeId: string): Promise<string>;
  calculateNice(pedigreeId: string): Promise<unknown>;
  getPedigreeSummary(pedigreeId: string): Promise<PedigreeSummary>;
}

export class EvageneClient implements EvageneApi {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly http: HttpGateway;

  constructor(options: EvageneClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/$/, '');
    this.apiKey = options.apiKey;
    this.http = options.http;
  }

  async fetchEmbedSvg(pedigreeId: string): Promise<string> {
    const path = `/api/embed/${pedigreeId}/svg`;
    const response = await this.http.getText(this.urlFor(path), this.authHeaders());
    ensureOk(response.status, 'GET', path);
    return response.text();
  }

  async calculateNice(pedigreeId: string): Promise<unknown> {
    const path = `/api/pedigrees/${pedigreeId}/risk/calculate`;
    const response = await this.http.postJson(
      this.urlFor(path),
      this.authHeaders({ 'Content-Type': 'application/json', Accept: 'application/json' }),
      { model: 'NICE' },
    );
    ensureOk(response.status, 'POST', path);
    return response.json();
  }

  async getPedigreeSummary(pedigreeId: string): Promise<PedigreeSummary> {
    const path = `/api/pedigrees/${pedigreeId}`;
    const response = await this.http.getText(
      this.urlFor(path),
      this.authHeaders({ Accept: 'application/json' }),
    );
    ensureOk(response.status, 'GET', path);
    const parsed = parseJsonObject(await response.text(), path);
    return {
      id: readString(parsed, 'id', path),
      displayName: readString(parsed, 'display_name', path),
    };
  }

  private urlFor(path: string): string {
    return `${this.baseUrl}${path}`;
  }

  private authHeaders(extra: Readonly<Record<string, string>> = {}): Record<string, string> {
    return { 'X-API-Key': this.apiKey, ...extra };
  }
}

function ensureOk(status: number, method: string, path: string): void {
  if (status < HTTP_OK_LOWER || status >= HTTP_OK_UPPER) {
    throw new EvageneApiError(
      `Evagene API returned HTTP ${status.toString()} for ${method} ${path}`,
    );
  }
}

function parseJsonObject(text: string, path: string): Record<string, unknown> {
  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch (error) {
    const reason = error instanceof Error ? error.message : String(error);
    throw new EvageneApiError(`Evagene API returned non-JSON body for ${path}: ${reason}`);
  }
  if (typeof parsed !== 'object' || parsed === null || Array.isArray(parsed)) {
    throw new EvageneApiError(`Evagene API returned non-object JSON for ${path}`);
  }
  return parsed as Record<string, unknown>;
}

function readString(container: Record<string, unknown>, key: string, path: string): string {
  const value = container[key];
  if (typeof value !== 'string') {
    throw new EvageneApiError(`Evagene API response for ${path} is missing string field '${key}'`);
  }
  return value;
}

File: tumour-board-briefing/python/src/tumour_board_briefing/evagene_client.py
"""Typed Evagene API client for the three endpoints the briefing needs.

One class, three methods:
- ``fetch_pedigree_detail`` — ``GET /api/pedigrees/{id}``
- ``fetch_pedigree_svg``    — ``GET /api/pedigrees/{id}/export.svg``
- ``calculate_risk``        — ``POST /api/pedigrees/{id}/risk/calculate``

Each method maps transport failures onto :class:`ApiError`; schema
interpretation is the aggregator's job, not the client's.
"""

from __future__ import annotations

from typing import Any

from .http_gateway import HttpGateway

_HTTP_OK_LOWER = 200
_HTTP_OK_UPPER = 300


class ApiError(RuntimeError):
    """Raised when the Evagene API is unreachable or returns a non-2xx response."""


class EvageneClient:
    def __init__(
        self,
        *,
        base_url: str,
        api_key: str,
        http: HttpGateway,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._http = http

    def fetch_pedigree_detail(self, pedigree_id: str) -> dict[str, Any]:
        url = f"{self._base_url}/api/pedigrees/{pedigree_id}"
        response = self._http.get(url, headers=self._json_headers())
        self._require_2xx(response, url)
        payload = response.json()
        if not isinstance(payload, dict):
            raise ApiError(
                f"Pedigree detail endpoint returned non-object JSON: {type(payload).__name__}"
            )
        return payload

    def fetch_pedigree_svg(self, pedigree_id: str) -> bytes:
        url = f"{self._base_url}/api/pedigrees/{pedigree_id}/export.svg"
        response = self._http.get(url, headers=self._svg_headers())
        self._require_2xx(response, url)
        return response.content

    def calculate_risk(
        self,
        pedigree_id: str,
        model: str,
        *,
        counselee_id: str | None = None,
    ) -> dict[str, Any]:
        url = f"{self._base_url}/api/pedigrees/{pedigree_id}/risk/calculate"
        body: dict[str, Any] = {"model": model}
        if counselee_id is not None:
            body["counselee_id"] = counselee_id

        response = self._http.post_json(url, headers=self._json_headers(), body=body)
        self._require_2xx(response, f"{url} (model={model})")
        payload = response.json()
        if not isinstance(payload, dict):
            raise ApiError(
                f"Risk endpoint returned non-object JSON for model {model}: "
                f"{type(payload).__name__}"
            )
        return payload

    def _require_2xx(self, response: Any, context: str) -> None:
        if not _HTTP_OK_LOWER <= response.status_code < _HTTP_OK_UPPER:
            raise ApiError(f"Evagene API returned HTTP {response.status_code} for {context}")

    def _json_headers(self) -> dict[str, str]:
        return {
            "X-API-Key": self._api_key,
            "Content-Type": "application/json",
            "Accept": "application/json",
        }

    def _svg_headers(self) -> dict[str, str]:
        return {
            "X-API-Key": self._api_key,
            "Accept": "image/svg+xml",
        }
